File: tools/poc/001-crt-effects-test/cpp/crt_test.cpp
// CRT Effects Test - GPU/CPU Parity Verification
// Standalone test that renders all profiles at multiple resolutions
// and compares GPU vs CPU output

#include <QApplication>
#include <QImage>
#include <QDir>
#include <QDebug>
#include <QOpenGLWidget>
#include <QOpenGLFunctions>
#include <QOpenGLShaderProgram>
#include <QOpenGLTexture>
#include <QOffscreenSurface>
#include <QOpenGLContext>
#include <QOpenGLFramebufferObject>

#include <cmath>
#include <vector>
#include <string>

// Include the actual CRT implementations
#include "widgets/crtfilter.h"
#include "widgets/crtprofiles.h"

// Test configuration
struct TestConfig {
    std::string name;
    CRTProfile profile;
    int outputWidth;
    int outputHeight;
};

// Reference ZX Spectrum screen dimensions
constexpr int SRC_WIDTH = 352;
constexpr int SRC_HEIGHT = 288;

// Generate a test pattern image (ZX Spectrum-like)
QImage generateTestPattern() {
    QImage img(SRC_WIDTH, SRC_HEIGHT, QImage::Format_RGBA8888);
    img.fill(QColor(192, 192, 192));  // Light gray background (ZX paper)

    // Draw some colored bars and text-like patterns
    QPainter p(&img);

    // Border area (darker gray)
    p.fillRect(0, 0, SRC_WIDTH, 48, QColor(192, 192, 192));
    p.fillRect(0, SRC_HEIGHT - 56, SRC_WIDTH, 56, QColor(192, 192, 192));

    // Simulated menu box
    int boxX = 140, boxY = 80, boxW = 150, boxH = 120;
    p.fillRect(boxX, boxY, boxW, boxH, QColor(0, 0, 0));
    p.fillRect(boxX + 2, boxY + 2, boxW - 4, boxH - 4, QColor(0, 255, 255));  // Cyan

    // Color bars (rainbow stripe)
    int stripeY = boxY + 8;
    int stripeW = 20;
    QColor colors[] = {Qt::black, Qt::red, Qt::yellow, Qt::green, Qt::cyan, Qt::blue, Qt::magenta};
    for (int i = 0; i < 7; i++) {
        p.fillRect(boxX + 8 + i * stripeW, stripeY, stripeW, 8, colors[i]);
    }

    // Simulated text lines
    p.setPen(Qt::black);
    p.setFont(QFont("Courier", 8));
    p.drawText(boxX + 8, boxY + 32, "Tape Loader");
    p.drawText(boxX + 8, boxY + 48, "128 BASIC");
    p.drawText(boxX + 8, boxY + 64, "Calculator");
    p.drawText(boxX + 8, boxY + 80, "48 BASIC");
    p.drawText(boxX + 8, boxY + 96, "TR-DOS");

    // Copyright text
    p.drawText(80, SRC_HEIGHT - 24, "(c) 1986 Sinclair Research Ltd");

    p.end();
    return img;
}

// Apply CPU CRT filter
QImage applyCpuCRT(const QImage& src, const CRTProfileParams& params, int outWidth, int outHeight) {
    // Scale source to output size
    QImage scaled = src.scaled(outWidth, outHeight, Qt::IgnoreAspectRatio, Qt::FastTransformation);
    scaled = scaled.convertToFormat(QImage::Format_RGBA8888);

    QImage result(outWidth, outHeight, QImage::Format_RGBA8888);

    CRTFilter filter;
    filter.apply(scaled.bits(), result.bits(), outWidth, outHeight,
                 src.width(), src.height(), params);

    return result;
}

// Calculate image statistics
struct ImageStats {
    double avgBrightness;
    double minBrightness;
    double maxBrightness;
    double stdDev;
};

ImageStats calculateStats(const QImage& img) {
    ImageStats stats = {0, 255, 0, 0};

    double sum = 0;
    double sumSq = 0;
    int count = 0;

    for (int y = 0; y < img.height(); y++) {
        const uchar* row = img.constScanLine(y);
        for (int x = 0; x < img.width(); x++) {
            // Calculate luminance
            double r = row[x * 4 + 0];
            double g = row[x * 4 + 1];
            double b = row[x * 4 + 2];
            double lum = 0.299 * r + 0.587 * g + 0.114 * b;

            sum += lum;
            sumSq += lum * lum;
            stats.minBrightness = std::min(stats.minBrightness, lum);
            stats.maxBrightness = std::max(stats.maxBrightness, lum);
            count++;
        }
    }

    stats.avgBrightness = sum / count;
    stats.stdDev = std::sqrt(sumSq / count - stats.avgBrightness * stats.avgBrightness);

    return stats;
}

// Compare two images
struct CompareResult {
    double mse;           // Mean squared error
    double psnr;          // Peak signal-to-noise ratio
    double avgDiff;       // Average absolute difference
    double maxDiff;       // Maximum difference
    bool similar;         // Within acceptable tolerance
};

CompareResult compareImages(const QImage& a, const QImage& b) {
    CompareResult result = {0, 0, 0, 0, false};

    if (a.size() != b.size()) {
        qWarning() << "Image sizes don't match!";
        return result;
    }

    double sumSqErr = 0;
    double sumAbsDiff = 0;
    double maxDiff = 0;
    int count = 0;

    for (int y = 0; y < a.height(); y++) {
        const uchar* rowA = a.constScanLine(y);
        const uchar* rowB = b.constScanLine(y);
        for (int x = 0; x < a.width() * 4; x++) {
            double diff = std::abs(static_cast<double>(rowA[x]) - rowB[x]);
            sumSqErr += diff * diff;
            sumAbsDiff += diff;
            maxDiff = std::max(maxDiff, diff);
            count++;
        }
    }

    result.mse = sumSqErr / count;
    result.psnr = (result.mse > 0) ? 10 * std::log10(255.0 * 255.0 / result.mse) : 100;
    result.avgDiff = sumAbsDiff / count;
    result.maxDiff = maxDiff;
    result.similar = (result.avgDiff < 5.0 && result.maxDiff < 30.0);

    return result;
}

int main(int argc, char* argv[]) {
    QApplication app(argc, argv);

    // Create output directory
    QString outputDir = "output";
    QDir().mkpath(outputDir);

    // Generate test pattern
    QImage testPattern = generateTestPattern();
    testPattern.save(outputDir + "/00_reference.png");
    qDebug() << "Generated reference image:" << SRC_WIDTH << "x" << SRC_HEIGHT;

    // Test configurations
    std::vector<TestConfig> tests;

    // All profiles
    std::vector<std::pair<std::string, CRTProfile>> profiles = {
        {"None", CRTProfile::None},
        {"Basic", CRTProfile::Basic},
        {"Aperture", CRTProfile::Aperture},
        {"ShadowMask", CRTProfile::ShadowMask},
        {"SlotMask", CRTProfile::SlotMask},
        {"Megatron", CRTProfile::Megatron}
    };

    // All resolutions (scale factors)
    std::vector<std::pair<std::string, float>> scales = {
        {"1x", 1.0f},
        {"1.5x", 1.5f},
        {"2x", 2.0f},
        {"2.5x", 2.5f},
        {"3x", 3.0f},
        {"4x", 4.0f}
    };

    qDebug() << "\n=== CRT Effects Test Suite ===\n";
    qDebug() << "Testing" << profiles.size() << "profiles at" << scales.size() << "resolutions\n";

    // Results summary
    int passed = 0, failed = 0;

    for (const auto& [profileName, profile] : profiles) {
        CRTProfileParams params = CRTProfileParams::FromProfile(profile);

        qDebug() << "\n--- Profile:" << profileName.c_str() << "---";
        qDebug() << "  maskType:" << static_cast<int>(params.maskType)
                 << "maskStrength:" << params.maskStrength
                 << "scanlineWeight:" << params.scanlineWeight;

        for (const auto& [scaleName, scale] : scales) {
            int outW = static_cast<int>(SRC_WIDTH * scale);
            int outH = static_cast<int>(SRC_HEIGHT * scale);

            // Apply CPU CRT
            QImage cpuResult = applyCpuCRT(testPattern, params, outW, outH);

            // Calculate stats
            ImageStats stats = calculateStats(cpuResult);

            // Save output
            QString filename = QString("%1/cpu_%2_%3.png")
                .arg(outputDir)
                .arg(QString::fromStdString(profileName))
                .arg(QString::fromStdString(scaleName));
            cpuResult.save(filename);

            // Report
            QString status;
            bool ok = true;

            // Check for burning white (avg brightness > 250)
            if (stats.avgBrightness > 250 && profile != CRTProfile::None) {
                status = "FAIL: Burning white!";
                ok = false;
            }
            // Check for too dark (avg brightness < 50)
            else if (stats.avgBrightness < 50 && profile != CRTProfile::None) {
                status = "FAIL: Too dark!";
                ok = false;
            }
            // Check None profile unchanged
            else if (profile == CRTProfile::None) {
                ImageStats refStats = calculateStats(testPattern.scaled(outW, outH));
                if (std::abs(stats.avgBrightness - refStats.avgBrightness) > 5) {
                    status = "FAIL: None profile modified image!";
                    ok = false;
                } else {
                    status = "OK";
                }
            }
            else {
                status = "OK";
            }

            if (ok) passed++; else failed++;

            qDebug().nospace()
                << "  " << scaleName.c_str() << " (" << outW << "x" << outH << "): "
                << "avg=" << QString::number(stats.avgBrightness, 'f', 1)
                << " min=" << QString::number(stats.minBrightness, 'f', 1)
                << " max=" << QString::number(stats.maxBrightness, 'f', 1)
                << " " << status;
        }
    }

    qDebug() << "\n=== Summary ===";
    qDebug() << "Passed:" << passed << " Failed:" << failed;
    qDebug() << "Output saved to:" << QDir::currentPath() + "/" + outputDir;

    return (failed == 0) ? 0 : 1;
}

File: tools/poc/001-crt-effects-test/python/crt_test.py
#!/usr/bin/env python3
"""
CRT Effects Test Suite
Tests all CRT profiles at multiple resolutions by analyzing output images.

This script generates test reference data and validates CRT filter behavior.
"""

import os
import sys
import json
from pathlib import Path
from dataclasses import dataclass
from typing import List, Tuple
import subprocess

# Try to import PIL for image analysis
try:
    from PIL import Image
    import numpy as np
    HAS_PIL = True
except ImportError:
    HAS_PIL = False
    print("Warning: PIL/numpy not available. Install with: pip install pillow numpy")

@dataclass
class ProfileConfig:
    name: str
    mask_type: int  # 0=None, 1=Aperture, 2=Shadow, 3=Slot
    mask_strength: float
    scanline_weight: float
    brightness: float
    contrast: float
    saturation: float

# CRT Profiles as defined in crtprofiles.cpp
PROFILES = [
    ProfileConfig("None", 0, 0.0, 0.0, 1.0, 1.0, 1.0),
    ProfileConfig("Basic", 0, 0.0, 0.3, 1.0, 1.0, 1.0),
    ProfileConfig("Aperture", 1, 0.5, 0.25, 1.0, 1.0, 1.1),
    ProfileConfig("ShadowMask", 2, 0.6, 0.2, 1.0, 1.0, 1.05),
    ProfileConfig("SlotMask", 3, 0.55, 0.35, 1.1, 1.0, 1.15),
    ProfileConfig("Megatron", 1, 0.7, 0.0, 1.05, 1.1, 1.2),
]

# Test resolutions (as scale factors from 352x288)
SCALES = [1.0, 1.5, 2.0, 2.5, 3.0, 4.0]

# Reference dimensions
SRC_WIDTH = 352
SRC_HEIGHT = 288


def calculate_expected_brightness(profile: ProfileConfig, scale: float) -> Tuple[float, float]:
    """
    Calculate expected brightness range for a profile at given scale.
    Returns (min_expected, max_expected) as 0-255 values.
    """
    base = 192  # ZX Spectrum gray background

    if profile.name == "None":
        return (base - 5, base + 5)

    # Apply brightness/contrast
    adjusted = base * profile.brightness

    # Apply scanline darkening (avg ~50% of lines affected)
    if profile.scanline_weight > 0:
        scanline_factor = 1.0 - profile.scanline_weight * 0.5
        adjusted *= scanline_factor

    # Apply mask darkening based on scale
    if profile.mask_strength > 0:
        if scale >= 3.0:
            # Full mask - average ~= 1 - 0.55 * strength
            mask_factor = 1.0 - 0.55 * profile.mask_strength
        elif scale >= 2.0:
            # Blended mask
            blend = (scale - 2.0)  # 0 at 2x, 1 at 3x
            full_mask = 1.0 - 0.55 * profile.mask_strength
            mask_factor = 1.0 - (1.0 - full_mask) * blend
        else:
            # Uniform compensation
            mask_factor = 1.0 - 0.55 * profile.mask_strength
        adjusted *= mask_factor

    # Allow 15% tolerance
    min_val = max(0, adjusted * 0.85)
    max_val = min(255, adjusted * 1.15)

    return (min_val, max_val)


def analyze_image(img_path: str) -> dict:
    """Analyze an image and return statistics."""
    if not HAS_PIL:
        return {"error": "PIL not available"}

    img = Image.open(img_path).convert('RGB')
    arr = np.array(img)

    # Calculate luminance
    lum = 0.299 * arr[:,:,0] + 0.587 * arr[:,:,1] + 0.114 * arr[:,:,2]

    return {
        "width": img.width,
        "height": img.height,
        "avg_brightness": float(np.mean(lum)),
        "min_brightness": float(np.min(lum)),
        "max_brightness": float(np.max(lum)),
        "std_brightness": float(np.std(lum)),
    }


def generate_test_image(output_path: str):
    """Generate a ZX Spectrum-like test pattern."""
    if not HAS_PIL:
        print("Cannot generate test image without PIL")
        return False

    img = Image.new('RGB', (SRC_WIDTH, SRC_HEIGHT), (192, 192, 192))

    # Simple pattern - just save for now
    # In a real test we'd draw the full ZX boot screen
    from PIL import ImageDraw
    draw = ImageDraw.Draw(img)

    # Draw a box
    draw.rectangle([140, 80, 290, 200], fill=(0, 255, 255), outline=(0, 0, 0))

    # Color bars
    colors = [(0,0,0), (255,0,0), (255,255,0), (0,255,0), (0,255,255), (0,0,255), (255,0,255)]
    for i, color in enumerate(colors):
        draw.rectangle([148 + i*20, 88, 168 + i*20, 96], fill=color)

    img.save(output_path)
    return True


def run_tests():
    """Run all CRT effect tests."""
    output_dir = Path(__file__).parent / "output"
    output_dir.mkdir(exist_ok=True)

    print("=" * 60)
    print("CRT Effects Test Suite")
    print("=" * 60)
    print()

    # Generate reference image
    ref_path = output_dir / "reference.png"
    if not ref_path.exists():
        print("Generating reference test image...")
        if not generate_test_image(str(ref_path)):
            print("Failed to generate reference image")
            return 1

    results = []
    passed = 0
    failed = 0

    for profile in PROFILES:
        print(f"\n--- Profile: {profile.name} ---")
        print(f"    mask_type={profile.mask_type} mask_strength={profile.mask_strength}")
        print(f"    scanline_weight={profile.scanline_weight}")

        for scale in SCALES:
            out_w = int(SRC_WIDTH * scale)
            out_h = int(SRC_HEIGHT * scale)

            # Calculate expected brightness
            exp_min, exp_max = calculate_expected_brightness(profile, scale)

            # For now, just print expected values
            # In full test, would capture actual screenshots
            print(f"    {scale}x ({out_w}x{out_h}): "
                  f"expected brightness {exp_min:.1f}-{exp_max:.1f}")

            result = {
                "profile": profile.name,
                "scale": scale,
                "resolution": f"{out_w}x{out_h}",
                "expected_min": exp_min,
                "expected_max": exp_max,
            }
            results.append(result)

    # Save results
    results_path = output_dir / "expected_values.json"
    with open(results_path, 'w') as f:
        json.dump(results, f, indent=2)
    print(f"\nExpected values saved to: {results_path}")

    # Print key findings
    print("\n" + "=" * 60)
    print("KEY VALIDATION POINTS:")
    print("=" * 60)
    print()
    print("1. Megatron at 1x scale:")
    print(f"   Expected brightness: {calculate_expected_brightness(PROFILES[5], 1.0)}")
    print("   Should show uniform gray, NOT burning white")
    print()
    print("2. Megatron at 3x scale:")
    print(f"   Expected brightness: {calculate_expected_brightness(PROFILES[5], 3.0)}")
    print("   Should show visible mask pattern")
    print()
    print("3. None profile at any scale:")
    print("   Should be ~192 (unchanged from source)")
    print()

    return 0


if __name__ == "__main__":
    sys.exit(run_tests())
